### PR TITLE
feat: implemented read_tool backend

### PR DESCRIPTION
**Pull request by EllaLi86**
Implemented backend file of reading tool for LLM integration

New API endpoint:
 -  `read_file`:  check the file path and file size, then read the file safely

Note: This PR only adds backend capabilities. Frontend wiring and LLM tool-calling logic (prompt + routing) will be handled in a separate PR

1. Start the Jupyter server.
2. Open JupyterLab and open the browser DevTools console.
3. Call the `read-file` endpoint directly from the console using fetch.
4. Test different cases by modifying `file_path` and `notebook_path` values.
[The XSRF token can be obtained from document.cookie.]

fetch('/dsc10-tutor-jlab-backend/read-file', {
  method: 'POST',
  headers: {
    'Content-Type': 'application/json',
    'X-XSRFToken': xsrfToken
  },
  credentials: 'same-origin',
  body: JSON.stringify({
    file_path: 'lab2.ipynb',
    notebook_path: 'example_notebooks/lab2/lab2.ipynb'
  })
})
.then(response => {
  console.log('Status:', response.status);
  if (!response.ok) {
    return response.json().then(err => Promise.reject(err));
  }
  return response.json();
})
.then(data => console.log('Success:', data))
.catch(error => console.error('Error:', error));

### CODE REVIEW

File: dsc10_tutor_jlab_backend/tests/test_handlers.py [hunk 1]
```diff
@@ -10,4 +14,88 @@ async def test_get_example(jp_fetch):
...
+            body=json.dumps({})
+        )
+    except HTTPClientError as e:
+        response = e.response
+
+    assert response.code == 400
+    payload = json.loads(response.body)
+    assert "error" in payload
+    assert "file_path" in payload["error"].lower()
+
```
Comment (abhi-sathvika) [inline dsc10_tutor_jlab_backend/tests/test_handlers.py]: Nice! The tests are very thorough, readable, and cover both happy paths and edge cases well!

File: dsc10_tutor_jlab_backend/handlers.py [hunk 2]
```diff
@@ -15,10 +22,149 @@ def get(self):
...
+        except Exception:
+            pass
+
+        return Path.cwd().resolve()
+
+    def _resolve_and_validate_path(self, file_path: str, safe_base_dir: Path) -> Path | None:
+        """
+        Resolve the file path and validate it's within the safe base directory.
+        Returns the resolved Path if valid, None otherwise.
+        """
```
Comment (abhi-sathvika) [inline dsc10_tutor_jlab_backend/handlers.py]: Nice! The resolve + relative_to approach for path validation is solid and avoids common path traversal pitfalls...making it more robust

File: src/api/index.ts [hunk 3]
```diff
@@ -171,3 +171,33 @@ export async function requestAPI<T>(
...
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json'
+    },
+    body: JSON.stringify({
+      file_path,
+      notebook_path
+    })
+  });
+}
```
Comment (abhi-sathvika) [inline src/api/index.ts]: we should consider moving readFile into a separate "file_tools_calls.ts" file to keep index.ts from growing too large and crowded as we add more tools

File: dsc10_tutor_jlab_backend/handlers.py [hunk 4]
```diff
@@ -15,10 +22,149 @@ def get(self):
...
+        try:
+            data = json.loads(self.request.body.decode('utf-8'))
+            file_path = data.get('file_path', '')
+
+            if not file_path:
+                self.set_status(400)
+                self.finish(json.dumps({
+                    "error": "file_path parameter is required"
+                }))
+                return
```
Comment (abhi-sathvika) [inline dsc10_tutor_jlab_backend/handlers.py]: To keep things consistent with the list_files & search_files PR, could we refactor the file-reading logic so that the handler logic lives in a separate module (e.g. file_tools.py) rather than directly in handlers.py? This would help keep handlers.py focused on routing/wiring and make the file-related logic a bit cleaner and easier to extend

**Review by abhi-sathvika — CHANGES_REQUESTED** [review]
Great work! Let me know if you have any questions about the comments

File: dsc10_tutor_jlab_backend/handlers.py [hunk 5]
```diff
@@ -1,9 +1,16 @@
...
 
 from jupyter_server.base.handlers import APIHandler
 from jupyter_server.utils import url_path_join
 import tornado
 
+
+MAX_FILE_SIZE = 10 * 1024 * 1024
+MAX_CONTENT_LENGTH = 1024 * 1024
+
+
```
[the remark this answers is not in the document]
Comment (EllaLi86) [reply dsc10_tutor_jlab_backend/handlers.py]: Got it! MAX_FILE_SIZE = 10 * 1024 * 1024 is 10 MB  and it represents how large the size of the file we're willing to read. MAX_CONTENT_LENGTH = 1024 * 1024 is 1 MB and it represents how much of that content we put in the JSON body. 
I chose these limits mainly to prevent reading overly large files into memory, which could impact performance. I was concerned that without these safeguards, opening huge files could take a long time. In order to test it and make it smoothly when I first tried, I just set these two numbers that large enough. That said, I’m not sure whether this concern is overkill — I’d love to hear the team’s opinion on whether we actually need to consider this in production, or if it’s too restrictive. It was like a double safeguard. If it doesn’t work in some cases, we can always adjust it later.

File: dsc10_tutor_jlab_backend/handlers.py [hunk 6]
```diff
@@ -1,9 +1,16 @@
 import json
+import os
+from pathlib import Path
 
 from jupyter_server.base.handlers import APIHandler
 from jupyter_server.utils import url_path_join
 import tornado
 
+
+MAX_FILE_SIZE = 10 * 1024 * 1024
```
[the remark this answers is not in the document]
Comment (EllaLi86) [reply dsc10_tutor_jlab_backend/handlers.py]: MAX_FILE_SIZE = 10 * 1024 * 1024 is 10 MB  and it represents how large the size of the file we're willing to read. MAX_CONTENT_LENGTH = 1024 * 1024 is 1 MB and it represents how much of that content we put in the JSON body. 
I chose these limits mainly to prevent reading overly large files into memory, which could impact performance. I was concerned that without these safeguards, opening huge files could take a long time. In order to test it and make it smoothly when I first tried, I just set these two numbers that large enough. That said, I’m not sure whether this concern is overkill — I’d love to hear the team’s opinion on whether we actually need to consider this in production, or if it’s too restrictive. It was like a double safeguard. If it doesn’t work in some cases, we can always adjust it.

Comment (preethim111) [inline dsc10_tutor_jlab_backend/handlers.py]: For very large files, could we maybe use streaming or chunked reads?
